### ISSUE TITLE
Order fill events notified before order creation events

### ISSUE THREAD

**Issue by abitmore** [bitshares/bitshares-core#1277]
**Bug Description**
> the market subscription thread seems to be guaranteed to fire before the transaction callback thread, which doesn't make sense for clients because it implies that the market order went through before it actually did.

A client that subscribed to a market always receives order fill events before order creation events if the orders are filled in a same block.

There is a `FIXME` in code: https://github.com/bitshares/bitshares-core/blob/d94cc574b2ca460dd636e573f25f0c7737b4a355/libraries/app/database_api.cpp#L2311-L2313

Probably because `on_applied_block()` is called before `on_objects_changed()`: https://github.com/bitshares/bitshares-core/blob/d94cc574b2ca460dd636e573f25f0c7737b4a355/libraries/chain/db_block.cpp#L540-L544

**Expected Behavior**
When a new order is placed, clients receive an order creation event first, then an order fill event.
 
**Additional Context (optional)**
I'm not sure if it's safe to simply move `on_objects_changed()` to be called before `on_applied_block()`.

**Impacts**
Describe which portion(s) of BitShares Core may be impacted by this bug. Please tick at least one box.
- [x] API (the application programming interface)
- [ ] Build (the build process or something prior to compiled code)
- [ ] CLI (the command line wallet)
- [ ] Deployment (the deployment process after building such as Docker, Travis, etc.)
- [ ] DEX (the Decentralized EXchange, market engine, etc.)
- [ ] P2P (the peer-to-peer network for transaction/block propagation)
- [ ] Performance (system or user efficiency, etc.)
- [ ] Protocol (the blockchain logic, consensus, validation, etc.)
- [ ] Security (the security of system or user data, etc.)
- [ ] UX (the User Experience)
- [ ] Other (please add below)

## CORE TEAM TASK LIST
- [ ] Evaluate / Prioritize Bug Report
- [ ] Refine User Stories / Requirements
- [ ] Define Test Cases
- [ ] Design / Develop Solution
- [ ] Perform QA/Testing
- [ ] Update Documentation

**Comment by arminnaderi** [comment]
Note that this problem also occurs with fill orders. If an order fills instantly on creation then the `market subscription` callback is called first and then the transaction is processed.

**Comment by arminnaderi** [comment]
Actually I misread the issue above. The problem you described is concerning the fill orders. The original problem we discussed in earlier chats was about the order creation notification (ie `subscribe_to_market` firing before the response of `broadcast_transaction_synchronous` is sent back to the client.

For example, the output of:

`"{ 'id': -1, 'method': 'call', 'params': [2, subscribe_to_market, [9001, "1.3.0", "1.3.21"]] }"`

is being returned to the client before the output of: 

```
"{ 'id': 1, 'method': 'call', 'params': [network_broadcast, broadcast_transaction_synchronous,[%s]]}"
```

**Comment by arminnaderi** [comment]
I think that if we consider the above bug^^ I mentioned to be not a problem, then its fine to leave the fill order bug as it currently is because the client can just listen for fill orders that do not have a matching order and put them in an "orphan" list, then when the order arrives the client can fill the order. It's a bit troublesome but I think if implemented well the code can run without any issues.

What are your thoughts @abitmore @pmconrad @elmato ?

**Comment by abitmore** [comment]
@arminnaderi actually I misunderstood your question earlier. Please create a new issue for that.

**Comment by arminnaderi** [comment]
Ok, will do now.

**Comment by abitmore** [comment]
Your issue is still related though. Notification of `broadcast_transaction_synchronous` is fired by `applied_block` signal (which is same to order fills but of different API container), notification of new orders is fired by `changed_objects` signal, there are threading involved so the order is not deterministic. In addition, because they're not in the same container (one is in `network_broadcast_api` and the other is in `database_api`), I think it's harder to fix.

**Comment by arminnaderi** [comment]
Don't think this issue can be fixed in a manner that doesn't cause side effect problems, closing

**Comment by abitmore** [comment]
It's still somewhat a bug. Although we're unable to fix it now, but perhaps can fix it in the future. So I'll keep this issue open.

**Comment by arminnaderi** [comment]
What I meant is order fill events and order creation events are not the **only** two events that are pushed by `on_applied_block()` and `on_objects_changed()`. Some clients might actually have no preference in the order of these events. However, I agree that order fill events should probably come after order creation events, if its fixed it will affect all other events that are pushed by these methods as a side effect

**Comment by ryanRfox** [comment]
Assigned to @OpenLedgerApp

**Comment by oxarbitrage** [comment]
I initialized some work in this issue however i didn't assigned myself and i didn't made a lot yet. 
Will like to know where @openledgerapp is up as i maybe can keep it or if they are advanced i can just abandon it.

**Comment by OpenLedgerApp** [comment]
We have already spent our hours on this issue. So, we want to keep it on us

**Comment by oxarbitrage** [comment]
ok, no worries.

**Comment by OpenLedgerApp** [comment]
Thanks!

**Comment by ryanRfox** [comment]
Sorry, @oxarbitrage I did not realize you already had some effort into this one. If you have anything to PR that may assist @OpenLedgerApp in their development, please do so. Thanks

**Comment by OpenLedgerApp** [comment]
Currently, from one side: Database_api subscribes to the events from DB (new, changed, removed) object. Filters and accumulates events related to (call_order_object, limit_order_object, force_settlement_object) and finally via fc::async notifies listeners.

From other side: Database_api subscribes to the event from DB (applied_block), Filters operations and accumulates events related to (fill_order_operation) and via fc::async notifies listeners.

There are some inconveniences:
1: Full match in the same block - Database_api emits only two fill_order events
2: Order of events from Database_api is not well defined as Database_api in fc::async use different queue for listeners.
3: Listeners receive different objects: *_order_oject, {fill_order_operation, fill_order_operation_result} or object_id

**This is draft version: and could be extended.**
We propose to extend Database_api:
Use only applied_block for filtering operations and accumulates events related to (limit_order_create_operation, fill_order_operation, limit_order_cancel_operation) and notifies listeners.
In this approach listeners will receive notifications regardless where they were happened (in the same block or not) and order will be defined.
See https://github.com/openledger/bitshares-core/commit/677a1517d2fb3f5be372f077ead80ce82ac7ad55

**Comment by arminnaderi** [comment]
I'm the original person who brought this issue up. It's not a problem anymore as it can be fixed client-side.

Websocket events are asynchronous and not guaranteed to come in any particular order.

@OpenLedgerApp Could you clarify what you mean by 1), "Full match in the same block" ?

**Comment by OpenLedgerApp** [comment]
create_sell_order(alice, core.amount(100), usd.amount(100));
create_sell_order(bob, usd.amount(100), core.amount(100));
If these operations are matched in the same block, listener will get only two fill_order notifications.

**Comment by arminnaderi** [comment]
@OpenLedgerApp You mean the listener will only send 2 fill_order notifications, and 0 order_creation (transaction callbacks) ?

I faced this problem when creating an order which was instantly filled in the same block. But the problem can be fixed client-side, it's not actually a technical bug.

Please check if you are receiving the transaction callbacks and paste your findings here

**Comment by OpenLedgerApp** [comment]
> You mean the listener will only send 2 fill_order notifications, and 0 order_creation (transaction callbacks) ?

Yes, database_api rases  2 fill_order notifications, and 0 order_creation.

**Comment by arminnaderi** [comment]
If the order was fully filled, no order_creation will be notified. This is normal behavior. Instead, a **full_order_object** will be sent to the websocket listener, which is basically a copy of the order

**Comment by OpenLedgerApp** [comment]
> If the order was fully filled the moment it was created, no order_creation will be notified. This is normal behavior. Instead, a **full_order_object** will be sent to the websocket listener, which is basically a copy of the order

Yes, the same logic I've mentioned above.
Are there any more requirements to #1277?

**Comment by arminnaderi** [comment]
Nope, that's why I mentioned there is no bug. You have all the information you need to properly handle the relevant events.

**Comment by arminnaderi** [comment]
Also, @OpenLedgerApp , please vote against the "Reduce MSSR of bitCNY to 1.02" poll worker if you want bitCNY to not go into global settlement. If a global settlement on an asset as crucial as bitCNY happens, many users will leave and it will hurt your bottom line.

**Comment by pmconrad** [comment]
No politics here please.

**Comment by abitmore** [comment]
@OpenLedgerApp expected behavior is described in OP:
> when a new order is placed, clients receive an order creation event first, then an order fill event.

**Comment by OpenLedgerApp** [comment]
> @OpenLedgerApp expected behavior is described in OP:
> 
> > when a new order is placed, clients receive an order creation event first, then an order fill event.
> 
> Take your use case as an example:
> 
> > **Use case**:
> > It is the case if orders are particularly matched in the different blocks:
> > steps:
> > 1: create_sell_order(alice, core.amount(100), usd.amount(100));
> > 2: wait some blocks
> > 3: create_sell_order(bob, usd.amount(200), core.amount(200));
> > result:
> > we have the following order of events:  **order_creation -> fill_order->fill_odrer -> order_creation** as it is mentioned #1277.
> > It is correct behavior! In the 1st step order is created and not matched. Then in the 3d step partly matching is happened (fill_order, fill_order ) and finally a new order was created (order_creation).
> 
> IMHO the correct order of events should be: order_creation(alice), order_creation(bob), fill_order(alice), fill_order(bob), no matter if alice's order or bob's has been fully filled or both.

@abitmore it could be done via operations. see https://github.com/openledger/bitshares-core/commit/677a1517d2fb3f5be372f077ead80ce82ac7ad55. You are welcome  to discuss it. Thanks.

**Comment by pmconrad** [comment]
We have two issues here:
1. there is no notification for order object creation if an incoming order is filled immediately
2. if an incoming order is partially filled, the notification for the fill will be sent before the notification for order object creation

The second point is a bug IMO, while the first may simply be unexpected but valid behaviour (if message order is not guaranteed like @arminnaderi said).

**Comment by arminnaderi** [comment]
@pmconrad I think the first issue is a bug from the perspective of writing clients, because you don't actually know if your order was created or not. Assuming that an order was created because a fill event pops up is too complicated.

Also, as I mentioned a few posts above, when the order is fully filled on creation, a **full_order_object** is sent to the websocket listener instead of the callback to **broadcast_transaction**

**Comment by OpenLedgerApp** [comment]
We have made PR - https://github.com/bitshares/bitshares-core/pull/1636

This PR describes a draft vision of market events model.

**Comment by jmjatlanta** [comment]
I have built some unit tests for creating markets and trades. This may help test this issue. I will update this comment when I find the relevant code.

**Comment by xloem** [comment]
A simple bandaid for this that might make life easier for the developers of client apps would be to include more order fields in the fill notification, notably the "for_sale" quantity, such that it could be used to initialize an object locally and to verify ordering of events.

**Comment by OpenLedgerApp** [comment]
> I have built some unit tests for creating markets and trades. This may help test this issue. I will update this comment when I find the relevant code.

When you can provide this test?

**Comment by jmjatlanta** [comment]
Sorry. I found it and then got distracted. Here it is:

https://github.com/bitshares/bitshares-core/pull/1661

Oh look! A squirrel....

**Comment by OpenLedgerApp** [comment]
> Sorry. I found it and then got distracted. Here it is:
> 
> #1661
> 
> Oh look! A squirrel....

Thanks.
Sure, this may help test this issue, but the use case, mentioned above, that reproduces the #1277 is not so complicated.
